### PR TITLE
[DSV4.1] Extend fused mHC combine/RMSNorm to wider verify batches

### PR DESCRIPTION
**Pull request by Stelath**
## Motivation

DeepSeek V4.1's fused four-stream combine/RMSNorm only handles up to eight rows. With eight requests and DSpark block size five, target verification can process 48 rows and falls back to separate combine and normalization kernels.

## Modifications

Extend the existing Blackwell BF16, width-5120 fast path to 96 rows. Use four output partitions through eight rows, two through 48, and one through 96 to reduce repeated statistics as row parallelism grows. Preserve the intermediate BF16 rounding, existing guards and fallback above 96 rows. No new flag or dependency.

Add 15 parameterized cases to the existing CUDA test covering partition boundaries, strided inputs and input scales. Independent draft against `dsv4.1` at `0ff9c986c`; no candidate-indexer dependency.

## Accuracy Tests

Prior GB300 validation of this implementation on SGLang `e087e662`: 216 seeded scale/stride cases; 197 of 49,274,880 values differed from native unfused combine/RMSNorm, by at most one BF16 ULP. All values were finite. The focused 15-case test also passed in our built image. This is not a bitwise-equivalence claim.

This draft rebases the unchanged kernel/dispatch change onto current `dsv4.1`. Syntax, diff whitespace, Ruff formatting and upstream-selected Ruff rules (`F401,F821,UP037`) pass locally. GPU execution on this new base and full upstream CI remain pending; no new GPU benchmark was run for this PR.

Reproduce: `pytest -q test/registered/kernels/ops/layernorm/test_hc_combine.py -k combine_norm_wide`.

## Speed Tests and Profiling

Prior same-GB300 CUDA-graph measurement at 48 rows: 4.267 us unfused versus 2.404 us fused (43.7% lower kernel latency), medians from three alternating-order trials. These are isolated operation timings, not end-to-end throughput gains. Current-base model quality and serving performance remain to be measured before merge.

## Checklist

- [ ] Full pre-commit hooks (targeted Ruff formatting/lint and syntax pass).
- [x] Extend an existing registered CUDA unit test.
- [x] No new user-facing option requiring documentation.
- [ ] Current-base accuracy and speed validation; prior measurements reported above.
- [x] Follow existing kernel and dispatch structure.

## Review and Merge Process

Draft for implementation review; GPU CI and current-base validation required before merge.

---
### CI States

Latest PR Test (Base): :x: [Run #34830024736](https://github.com/sgl-project/sglang/actions/runs/34830024736)
Latest PR Test (Extra): :x: [Run #34830024207](https://github.com/sgl-project/sglang/actions/runs/34830024207)
Latest PR Test (AMD ROCm 10): :x: [Run #34830024901](https://github.com/sgl-project/sglang/actions/runs/34830024901)

### CODE REVIEW

**Comment by sethforprivacy** [comment]
GB10 serving observation for `4c17445f15a721cd3ccbaa7ed665e12b83d0e02c`: our strict exact-output gate detects divergence after widening this dispatch. This is **not** a refutation of the PR's stated one-BF16-ULP bound or evidence of semantic-quality degradation.

Four DGX Sparks / GB10, TP4/EP4; SGLang `e087e662ba1ac4ef7747537e2a9141085efd4561` with the #39187 memory backport and #39068 in both arms. We then applied this PR's kernel/dispatch hunks. Model revision `dba1be0a40aa45a94ad051997016db3960a90277`, Mia adapter `e59e6eb67479aa68f6fa700c600dc90a0729b5ec`, packed NVMe Engram, chunk 4096, context 430080, requested KV 1500000, max running 8, DSpark block 5/static verify with natural acceptance. PyTorch 2.13.0+cu130, Triton 3.7.1, FlashInfer 0.6.18, b12x MXFP8; expandable segments and fused MoE finalize disabled. These are patched older-base tests, not execution of the whole current PR branch.

Our fixed gate sends each of 20 prompts at temperature 0 twice: max_tokens 1 with top_logprobs 20, then max_tokens 48. Sixteen prompts are short and four are longer inputs. The preceding memory+#39068 arm matched the fixed reference 20/20 for both first-token output and exact continuation. With this patch:

- First-token output matched 18/20.
- Exact continuation matched 13/20.
- All seven differing continuation cases were short inputs (13–21 actual prompt tokens); the four long-input cases matched.
- Seven-key auth and all seven vision/tool cases passed. Exact retrieval passed at 130258, 260119 and 406713 prompt tokens. Twelve decode samples had zero request errors.

One short fixture showing a first-token change:

```json
{"messages":[{"role":"user","content":"In one paragraph, describe how a sliding-window attention layer differs from full attention."}],"temperature":0,"max_tokens":1,"logprobs":true,"top_logprobs":20}
```

It tokenizes to 21 prompt tokens in this setup; the first output changed from `A` to `Full`. An IPv4-regex fixture also changed its first output from a single backtick to a fenced-code opener. We did not score these alternative answers as semantically wrong.

The widened row-count guard also covers short prefills, not only target verification, which is relevant to interpreting first-token changes. We have not isolated the numerical difference inside the operation on GB10, and the existing one-ULP evidence makes some generation divergence plausible. We rejected this candidate under our stricter exact-output requirement and restored the preceding profile, which passed both standard and extended qualification again. Reporting the observation to inform current-base/model-quality validation, rather than calling it corruption or requesting a revert. Throughput samples were noisy, so we are not making a performance claim for this arm.

**Comment by sethforprivacy** [comment]
Follow-up on the control's repeatability: a later reboot/requalification of the preceding memory + #39068, block-5 profile (without #39158) matched 20/20 first-token outputs but only 19/20 exact continuations. The one differing continuation was the same short IPv4-regex fixture: a fragment used `1\d\d` instead of `1\d{2}`. The preceding reboot of this profile had matched 20/20 continuations, and the two launch receipts are byte-identical. Auth, all seven vision/tool cases and exact retrieval through 406713 prompt tokens still passed; no semantic degradation has been established.

This introduces a control-repeatability confound for interpreting our earlier candidate comparisons. The earlier observations remain accurate, but continuation divergence alone cannot yet be causally attributed to a candidate patch. We have stopped additional tuning while the existing recovery sequence runs; isolating source/runtime state and repeated fixed-input behavior is needed before a stronger correctness conclusion. In particular, this is not evidence against #39158's stated numerical bound, nor does it establish that #39068 caused the variation.


Further diagnostic evidence: the recovered block-5 profile then produced identical text in 80 short C1 requests (four fixed fixtures, ten repeats each, with and without continuation logprobs). At the historical regex divergence position, the new passing capture reports exactly equal logprobs for `{` and `\d`: both -0.6931482553482056. These yield the equivalent fragments `1\d{2}` and `1\d\d`. This strengthens a tie-sensitive explanation for that particular intermittent continuation mismatch, but we did not capture this position's logits on the failing boot, so it does not establish its cause. It also does not explain away #39158's separate first-token distribution changes. Full current overlay hashes and block size match across all four ranks. This was a warmed, single-boot test; cross-boot repeatability remains unproven.
